src/components/map/EventPopup.js

```diff
@@ -0,0 +1,216 @@
+    });
+    const { programStageDataElements } = data;
+    let displayElements = [];
+    let eventCoordinateFieldName;
+
+    if (Array.isArray(programStageDataElements)) {
+        displayElements = programStageDataElements
+            .filter(d => d.displayInReports)
+            .map(d => d.dataElement);
+
+        for (let d of displayElements) {
+            await loadOptionSet(d);
+        }
+
+        if (eventCoordinateField) {
+            const coordElement = programStageDataElements.find(
+                d => d.dataElement.id === eventCoordinateField
+            );
+
+            if (coordElement) {
+                eventCoordinateFieldName = coordElement.dataElement.name;
+            }
+        }
+    }
+
+    return { displayElements, eventCoordinateFieldName };
+};
+
+// Loads event data for the selected feature
+const loadEventData = async feature =>
+    feature ? apiFetch(`/events/${feature.id}`) : null;
+
+// Returns table rows for all display elements
+const getDataRows = (displayElements, dataValues, styleDataItem, value) => {
+    const dataRows = [];
+
+    // Include data element used for styling if not included below
+    if (
+        styleDataItem &&
+        !displayElements.find(d => d.id === styleDataItem.id)
+    ) {
+        dataRows.push(
+            <tr key={styleDataItem.id}>
+                <th>{styleDataItem.name}</th>
+                <td>{hasValue(value) ? value : i18n.t('Not set')}</td>
+            </tr>
+        );
+    }
+
+    // Include rows for each displayInReport data elements
+    displayElements.forEach(({ id, name, valueType, options }) => {
+        const { value } = dataValues.find(d => d.dataElement === id) || {};
+        let formattedValue = value;
+
+        if (valueType === 'COORDINATE' && value) {
+            formattedValue = formatCoordinate(value);
+        } else if (options) {
+            formattedValue = options[value];
+        } else if (!hasValue(value)) {
+            formattedValue = i18n.t('Not set');
+        }
+
+        dataRows.push(
+            <tr key={id}>
+                <th>{name}</th>
+                <td>{formattedValue}</td>
+            </tr>
+        );
+    });
+
+    if (dataRows.length) {
+        dataRows.push(<tr key="divider" style={{ height: 5 }} />);
+    }
+
+    return dataRows;
+};
+
+// Will display a popup for an event feature
+const EventPopup = props => {
+    const {
+        coordinates,
+        feature,
+        programStage,
+        eventCoordinateField,
+        styleDataItem,
+        onClose,
+    } = props;
+
+    const [eventData, setEventData] = useState();
+    const [dataElements, setDataElements] = useState({});
+    const { displayElements = [], eventCoordinateFieldName } = dataElements;
+
+    // Load data elements every time programStage is changed
+    useEffect(() => {
+        loadDataElements(programStage, eventCoordinateField).then(
+            setDataElements
+        );
+    }, [programStage]);
```

Comment:
Same as `loadEventData`, this needs to be able to abort.

Reply:
As this load should only happen once per layer (and not per popup), I've moved it back to the EventLayer component where I think it belongs.